Add support for OSDP Secure Channel

sidcha opened the pull request:
This patch adds Secure Channel capabilities to osdp Control Panel and
Peripheral Device modes.

Signed-off-by: Siddharth Chandrasekaran <siddharth@embedjournal.com>

cbsiddharth commented:
CC: @jukkar @carlescufi

jukkar commented on subsys/mgmt/osdp/src/osdp_cp.c:
```diff
@@ -729,117 +729,39 @@ int osdp_cp_set_callback_card_read(
 }
 
-int osdp_cp_send_cmd_output(int pd, struct osdp_cmd_output *p)
+int osdp_cp_send_command(int pd, struct osdp_cmd *p)
```
Use of these kind of short variable names like `p` here should be avoided (perhaps exception is the use of `i` in loops) as it does not give much hint for the usage of the variable. Not sure what it should be, perhaps containing `cmd` in the name.

cbsiddharth replied:
I will make this `struct osdp_cmd *cmd`. Thanks.

jukkar commented on subsys/mgmt/osdp/Kconfig:
```diff
@@ -56,6 +56,36 @@ config OSDP_PACKET_TRACE
+	select CRYPTO_MBEDTLS_SHIM
+	select MBEDTLS
+	select MBEDTLS_CIPHER_CCM_ENABLED
+	select TEST_RANDOM_GENERATOR
```
You should not select the test random generator, it is only meaningful in testing and should be only selected by sample prj.conf if needed.

cbsiddharth replied:
When user selects Secure Channel feature in menuconfig, there must be a random function defined and available for build to succeed. I was having it in my `prj.conf` file during development but moved it here for better usability. What is the right way to ask for randomness in Zephyr?

dleach02 replied:
@cbsiddharth, let the platform definition select the RANDOM_GENERATOR. If there isn't one then there isn't a true RNG support for the platform.

cbsiddharth replied:
@dleach02, Thank you. For now, I'll put `CONFIG_TEST_RANDOM_GENERATOR=y` in the samples's prj.conf.

dleach02 replied:
what platform are you testing on that causes you to add this?

cbsiddharth replied:
I use a STM32F1x (board/stm32_min_dev) for development and it does not have a TRNG.

cbsiddharth replied:
Thinking about it, shouldn't zephyr have a way to auto determine the most secure way to generate randomness based on platform capabilities?

This seems like a reasonable ask for me; but I'm just beginning to familiarizing myself with Zephyr.

dleach02 replied:
We have these types of discussions in the security working group. There are two types of random numbers being needed by an application; cryptographic and non-cryptographic. With non-cryptographic using the TEST_RANDOM_GENERATOR or a timer based one for seed into XOROSHIRO algorithm is fine. If the application/subsystem needs a cryptographically secure RNG value then it should use the csrand functions which will propagate down to only building with valid entropy sources (TRNG). If the STM32F1x doesn't have HW entropy then technically speaking it shouldn't be allowed to use cryptographic functions. 

Some of this groundwork has started with the addition of csrand based functions. 

https://docs.zephyrproject.org/latest/reference/random/index.html

cbsiddharth replied:
> We have these types of discussions in the security working group

Is this a slack channel or mailing list? I'd love to listen-in.

> If the STM32F1x doesn't have HW entropy then technically speaking it shouldn't be allowed to use cryptographic functions.

I agree. But shouldn't there be a fallback mechanism? I'd love to use the csrand methods, but still be able to do a spot check on a STM32F1x as I have hooked up 2 of them together to test both my CP and PD modes before publishing code upstream.

Basically, I feel a wrapper around `sys_rand_get` and `sys_csrand_get` based on platform capabilities (choosing the most secure option available) so application code can remain unchanged.

dleach02 replied:
The TEST_RANDOM_GENERATOR is for test cases to use and not a fallback for actual end products. There is a security slack channel to engage in as well.

> Basically, I feel a wrapper around sys_rand_get and sys_csrand_get based on platform capabilities (choosing the most secure option available) is needed so application code can remain unchanged.

Definitely not an option. You do not want to present a false sense of security on platforms that can't support it. The objective of the rand split was to start having improving security profile on applications and platforms built with Zephyr. If the platform doesn't have a cryptographic entropy source then it should not host any protocols or applications that claim to have security.

Now at some point someone might add support for a software based entropy generation but there has been no work on that so far. Most MCU devices have either a true hardware source for random values or a HW source to retrieve entropy values to seed a CSPRNG function.

cbsiddharth replied:
Okay.

jukkar commented on subsys/mgmt/osdp/Kconfig.cp:
```diff
@@ -41,3 +41,14 @@ config OSDP_PD_POLL_RATE
+
+config OSDP_MASTER_KEY
+	string "Secure Channel Master Key"
+	default "00112233445566778899aabbccddeeff"
```
This forces all the devices, that use this zephyr image, to have same master key. Is this really intended to work like this?

cbsiddharth replied:
This key must be unique to each device. This was a arbitrary default I added. Adding this to prj.conf also doesn't make a difference here. Perhaps this should come from devicetree?

cbsiddharth replied:
For now, I've moved it into the prj.conf where its more visible that each app must set this.

Doing this causes a harmless warning when OSDP mode is set to PD. I'll split the samples (later today) into CP and PD and it will go away.
```
warning: OSDP_MASTER_KEY (defined at subsys/mgmt/osdp/Kconfig.cp:47) was assigned the value
'11223344556677889911bbccddeeff' but got the value ''. Check these unsatisfied dependencies:
OSDP_MODE_CP (=n).
```

jukkar commented on subsys/mgmt/osdp/src/osdp.c:
```diff
@@ -199,7 +204,24 @@ static int osdp_init(const struct device *arg)
+		k_panic();
+	}
+
+#ifdef CONFIG_OSDP_SC_ENABLED
```
Better to have `if (IS_ENABLED(CONFIG_OSDP_SC_ENABLED)) {` here if possible

cbsiddharth replied:
Will address this in my next patchset.

jukkar commented on subsys/mgmt/osdp/src/osdp.c:
```diff
@@ -199,7 +204,24 @@ static int osdp_init(const struct device *arg)
+			LOG_ERR("Failed to parse key buffer");
+			return -1;
+		}
+		key = key_buf;
```
If we use the `IS_ENABLED()` like I suggested above, then this would not work any more as we would have a pointer to stack variable that is going away when exiting this `if` block. So the `key_buf` would be needed to move out of the `if` block in that case.

jukkar commented on subsys/mgmt/osdp/src/osdp_common.c:
```diff
@@ -74,3 +79,176 @@ struct osdp_cmd *osdp_cmd_get_last(struct osdp_pd *pd)
+
+#ifdef CONFIG_OSDP_SC_ENABLED
+
+static inline int hex2int(char ch)
```
There are already hex related functions like `hex2bin()` in `include/sys/util.h`, can't you use those?

cbsiddharth replied:
Thank you. I've switched to `hex2bin()` and removed my implementation.

jukkar commented on subsys/mgmt/osdp/src/osdp_common.c:
```diff
@@ -74,3 +79,176 @@ struct osdp_cmd *osdp_cmd_get_last(struct osdp_pd *pd)
+	struct osdp_pd *pd;
+	struct osdp *ctx = osdp_get_ctx();
+
+#ifdef CONFIG_OSDP_MODE_CP
```
Better would be to use `if (IS_ENABLED(....)) {` here if possible

cbsiddharth replied:
The condition check was unnecessary. Got rid of it.

jukkar commented on subsys/mgmt/osdp/src/osdp_cp.c:
```diff
@@ -627,6 +627,14 @@ static inline void cp_set_state(struct osdp_pd *pd, enum osdp_cp_state_e state)
 	CLEAR_FLAG(pd, PD_FLAG_AWAIT_RESP);
```
In commit `osdp/mgmt: CP: Reset channel before sending a command ` you probably meant to say `mgmt/osdp: CP: Reset channel before sending a command` in commit subject.

cbsiddharth replied:
Yes. Fixed, thanks.

sidcha committed:
mgmt/osdp: Add support for Secure Channel

This patch adds Secure Channel capabilities to osdp Control Panel and
Peripheral Device modes.

Signed-off-by: Siddharth Chandrasekaran <siddharth@embedjournal.com>

MaureenHelm commented:
@cbsiddharth is it possible to separate the bug fix from the new feature?

cbsiddharth commented:
> @cbsiddharth is it possible to separate the bug fix from the new feature?

This patchset began with initial changes for secure channel feature addition and then the feature itself. Now I see it does have quite a number of other changes too. I will try to split this PR as best as I can later this week.

dleach02 commented on subsys/mgmt/osdp/src/osdp_common.c:
```diff
@@ -74,3 +79,122 @@ struct osdp_cmd *osdp_cmd_get_last(struct osdp_pd *pd)
+
+void osdp_fill_random(uint8_t *buf, int len)
+{
+	sys_rand_get(buf, len);
```
re: conversation on random functions, if this use of random fill is a cryptographic usage then you should consider using sys_csrand_get().

cbsiddharth replied:
Yes. This is a cryptographic usage ([see my other comment](https://github.com/zephyrproject-rtos/zephyr/pull/28200#discussion_r489980793)).

dleach02 replied:
Ok. This should be a sys_csrand_get() call then.

cbsiddharth replied:
Just ordered a new board with STM32G474RE. I'll change this once I can try it out.

cbsiddharth replied:
Switched to `sys_csrand_get()`.

cbsiddharth commented:
I have addressed all comments so-far and awaiting review on this PR. Thanks!

d3zd3z commented on subsys/mgmt/osdp/src/osdp_sc.c:
```diff
@@ -0,0 +1,249 @@
+	const uint8_t *_s2 = s2;
+
+	for (i = 0; i < len; i++) {
+		if (_s1[i] != _s2[i]) {
```
This isn't actually constant time, as you conditionally increment based on whether the bytes match. You should do something like:
```C
for (i = 0; i < len; i++) {
        ret |= _s1[i] ^ _s2[i];
}
return ret;
```

cbsiddharth replied:
Very nice, thank you.
